### PR TITLE
Port to Manifest v3 (MV3)

### PR DESCRIPTION
**Pull request by michaelfig**
Allow building both Manifest v2 (compatible with both Firefox and Chrome) and Manifest v3 (Chrome-only).

Breaking change: `manifest.json` is split into `manifest-v2.json` and `manifest-v3.json`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,8 +1,13 @@
 {
 	"private": true,
 	"scripts": {
-		"build": "parcel build source/manifest.json --no-content-hash --no-source-maps --dist-dir distribution --no-cache --detailed-report 0",
-		"demo": "parcel serve demo/wallet/index.html demo/wallet-bridge.html --port=8000 --dist-dir demo-dist",
+    "clean:v2": "rm -rf distribution&&cp source/manifest-v2.json source/manifest.json",
+    "clean:v3": "rm -rf distribution&&cp source/manifest-v3.json source/manifest.json",
```
Comment (kriskowal) [inline package.json]: Super nit: spaces around &&

File: package.json [hunk 2]
```diff
@@ -1,8 +1,13 @@
...
+    "build": "parcel build source/manifest.json --no-content-hash --no-source-maps --dist-dir distribution --no-cache --detailed-report 0",
+    "demo": "run-p demo:*",
+    "demo:8000": "npm run serve -- --dist-dir demo-dist/8000 --port=8000",
+    "demo:5005": "npm run serve -- --dist-dir demo-dist/5005 --port=5005",
```
Comment (kriskowal) [inline package.json]: I’m inclined to favor names over port numbers for the script and the distribution directories, for transparency, if possible.

File: patches/@parcel+transformer-webextension+2.0.1.patch [hunk 3]
```diff
@@ -11,7 +11,42 @@ index fbc18aa..7b5e611 100644
...
-@@ -250,6 +250,10 @@ async function collectDependencies(asset, program, ptrs, hot) {
+@@ -194,16 +194,31 @@ async function collectDependencies(asset, program, ptrs, hot) {
+   if (program.web_accessible_resources) {
+     let war = [];
+ 
+-    for (let i = 0; i < program.web_accessible_resources.length; ++i) {
++    let i = 0;
++    for (const resource of program.web_accessible_resources) {
+       // TODO: this doesn't support Parcel resolution
+-      const globFiles = (await (0, _utils().glob)(_path().default.join(_path().default.dirname(filePath), program.web_accessible_resources[i]), fs, {})).map(fp => asset.addURLDependency(_path().default.relative(_path().default.dirname(filePath), fp), {
```
Comment (kriskowal) [inline patches/@parcel+transformer-webextension+2.0.1.patch:22]: Are we tracking this feature upstream?

File: source/browser-messages.js [hunk 4]
```diff
@@ -13,6 +13,11 @@ export const makeBrowserMessageHandler = ({
 }) => {
   return async obj => {
     switch (obj.type) {
+      case 'POWERBOX_INIT': {
+        // Tell our caller we're ready.
+        send({ type: 'POWERBOX_READY' });
```
Comment (kriskowal) [inline source/browser-messages.js:18]: Nice.

File: source/inject.js [hunk 5]
```diff
@@ -5,14 +5,15 @@ export const makeInjectScript = ({ document = globalThis.document }) => {
...
-  return content => {
+  return contentUrl => {
     const container = document.head || document.documentElement;
     const scriptTag = document.createElement('script');
-    scriptTag.setAttribute('async', false);
-    scriptTag.textContent = content;
+    scriptTag.src = contentUrl;
+    scriptTag.onload = function onLoad() {
+      this.remove();
+    };
```
Comment (kriskowal) [inline source/inject.js]: Always a nice touch.

File: source/manifest-v2.json [hunk 6]
```diff
@@ -26,5 +26,6 @@
     "run_at": "document_start",
     "js": ["content.js"],
     "all_frames": true
-  }]
+  }],
+  "web_accessible_resources": ["install-powerbox.js"]
```
Comment (kriskowal) [inline source/manifest-v2.json]: TIL, nice.

File: source/content.js [hunk 7]
```diff
@@ -19,7 +18,7 @@ const connect = makeConnect({ document, window });
 
 // If the injection throws, refuse to listen to messages.
 try {
-  injectScript(`globalThis.powerbox = (${createPowerboxInBrowser})(window);`);
+  injectScript(browser.runtime.getURL('/install-powerbox.js'));
```
Comment (kriskowal) [inline source/content.js:21]: Ah, also TIL.

File: source/manifest-v3.json [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+			"id": "powerbox@agoric.com",
+			"strict_min_version": "67.0"
+		}
+	},
+	"icons": {
+		"128": "icon.png"
+	},
+	"permissions": [
+		"storage",
+    "<all_urls>"
```
Comment (kriskowal) [inline source/manifest-v3.json:19]: Weird indent throughout this file.

File: source/manifest-v3.json [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+	"manifest_version": 3,
+	"minimum_chrome_version": "74",
+	"applications": {
+		"gecko": {
+			"id": "powerbox@agoric.com",
+			"strict_min_version": "67.0"
+		}
+	},
+	"icons": {
+		"128": "icon.png"
```
Comment (kriskowal) [inline source/manifest-v3.json]: I wrote up a script for icon generation. Is there already automation for that in any of the tools you’re using?

**Review by kriskowal — APPROVED** [review]

File: patches/@parcel+transformer-webextension+2.0.1.patch (hunk 3 above)
Comment (michaelfig) [reply patches/@parcel+transformer-webextension+2.0.1.patch:22]: > Are we tracking this feature upstream?

Yep, updated: https://github.com/parcel-bundler/parcel/issues/7458

File: source/manifest-v3.json (hunk 9 above)
Comment (michaelfig) [reply source/manifest-v3.json]: > I wrote up a script for icon generation. Is there already automation for that in any of the tools you’re using?

No, there isn't.  That would be a great contribution!

**Pull request merged**